software/controller/lib/hal/gpio.cpp

```diff
@@ -58,71 +36,111 @@ void enable_all_clocks() {
   enable_peripheral_clock(PeripheralID::GPIOH);
 }
 
-void pin_mode(Port port, uint8_t pin, PinMode mode) {
-  Register *const gpio = base_address(port);
-  gpio->mode &= ~(0b11 << (pin * 2));
-  gpio->mode |= (static_cast<uint32_t>(mode) << (pin * 2));
-}
+Pin::Pin(Port port, uint8_t pin, PinMode mode) : pin_(pin) {
+#if defined(BARE_STM32)
+  gpio_ = reinterpret_cast<RegisterStructure *>(port);
+  // reset the mode bits to 0
+  gpio_->mode &= ~(0b11 << (pin_ * 2));
+  // set the mode bits to the desired value
+  gpio_->mode |= (static_cast<uint32_t>(mode) << (pin_ * 2));
+#endif
+};
 
-void output_type(Port port, uint8_t pin, OutType output_type) {
-  Register *const gpio = base_address(port);
+#if defined(BARE_STM32)
```

Comment:
Really want to avoid having blocks of implementation be "ifdeffed". If this is a sizeable chunk of code, as it is here, maybe it is time to split this out into Abstract + Architecture-specific + Mock classes?